v2/health.go

```diff
@@ -0,0 +1,146 @@
+	// HealthCheckTimeout is the context timeout for each health check
+	HealthCheckTimeout time.Duration
+}
+
+// NewHealthChecker creates a new HealthChecker with the provided clients.
+// clients is a map of namespaceName name to azservicebus.Client.
+func NewHealthChecker(clients map[string]*azservicebus.Client, options *HealthCheckerOptions) *HealthChecker {
+	if options == nil {
+		options = &HealthCheckerOptions{}
+	}
+	if options.HealthCheckInterval <= 0 {
+		options.HealthCheckInterval = defaultHealthCheckInterval
+	}
+	if options.HealthCheckTimeout <= 0 || options.HealthCheckTimeout > options.HealthCheckInterval {
+		options.HealthCheckTimeout = options.HealthCheckInterval
+	}
+
+	return &HealthChecker{
+		clients: clients,
+		options: options,
+	}
+}
+
+// StartPeriodicHealthCheck starts the periodic health check for the provided HealthCheckable.
+// The health check will run on each client in the HealthChecker.
+// Stops when the context is canceled.
+func (h *HealthChecker) StartPeriodicHealthCheck(ctx context.Context, hc HealthCheckable) {
+	for namespaceName, client := range h.clients {
+		go func(namespaceName string, client *azservicebus.Client) {
+			h.periodicHealthCheck(ctx, hc, namespaceName, client)
+		}(namespaceName, client)
+	}
+}
+
+func (h *HealthChecker) periodicHealthCheck(ctx context.Context, hc HealthCheckable, namespaceName string, client *azservicebus.Client) {
+	nextCheck := time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Until(nextCheck)):
+			sbCtx, cancelFunc := context.WithTimeout(ctx, h.options.HealthCheckTimeout)
+			err := hc.HealthCheck(sbCtx, client, cancelFunc)
+			if err != nil {
+				getLogger(ctx).Error(err.Error())
+			}
+			hc.IncHealthCheckMetric(namespaceName, err)
+			nextCheck = nextCheck.Add(h.options.HealthCheckInterval)
+		}
+	}
+}
+
+// HealthCheckable is an interface for performing health checks on azservicebus.Sender and azservicebus.Receiver.
+type HealthCheckable interface {
+	HealthCheck(ctx context.Context, client *azservicebus.Client, cancelFunc context.CancelFunc) error
+	IncHealthCheckMetric(namespaceName string, healthCheckErr error)
+}
+
+// SenderHealthChecker performs health checks on azservicebus.Sender.
+type SenderHealthChecker struct {
+	EntityName string
+}
+
+// HealthCheck performs a health check on azservicebus.Sender by creating a new message batch.
+func (s *SenderHealthChecker) HealthCheck(ctx context.Context, client *azservicebus.Client, cancelFunc context.CancelFunc) error {
+	defer cancelFunc()
+	sbSender, err := client.NewSender(s.EntityName, nil)
+	if err != nil {
+		return err
+	}
+	if _, err = sbSender.NewMessageBatch(ctx, nil); err != nil {
+		return err
+	}
+	return sbSender.Close(ctx)
+}
+
+// IncHealthCheckMetric increments the sender health check metric based on the health check result.
+func (s *SenderHealthChecker) IncHealthCheckMetric(namespaceName string, healthCheckErr error) {
```

Comment:
I think this func should stay internal. unless you have a use-case for exporting it

Reply:
My original thought is this allows applications to define their own implementation of `HealthCheckable`. For example, if they want to use a different method to verify the connection, or if they want to increment some of their own metrics.

Also open to keeping it internal.